### PR TITLE
Use the interpreted program's TZ variable in localtime_r

### PR DESCRIPTION
**Pull request by saethlin**
This requires a bit of wiring and a new dependency, but the tests should correctly pass now regardless of what the host's time zone is.

Fixes https://github.com/rust-lang/miri/issues/3522

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -690,6 +712,15 @@ dependencies = [
  "windows-targets 0.48.5",
 ]
 
+[[package]]
+name = "parse-zoneinfo"
+version = "0.1.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "f4ee19a3656dadae35a33467f9714f1228dd34766dbe49e10e656b5296867aea"
+dependencies = [
+ "regex 0.2.11",
```
Comment (RalfJung) [inline Cargo.lock:721]: This is an ancient version of the regex crate -- is parse-zoneinfo even still maintained?

I don't think it would be good to bring such an ancient regex back into the rustc workspace.

Comment (RalfJung) [reply Cargo.lock:721]: Ah the main issue is that this version of parse-zoneinfo is from 6 years ago. There's version 0.3.0 from 4 years ago that uses a more recent `regex`.

I don't think we should add a dependency on regex 0.2. That's just too old.

File: Cargo.toml [hunk 2]
```diff
@@ -25,6 +25,7 @@ aes = { version = "0.8.3", features = ["hazmat"] }
 measureme = "11"
 ctrlc = "3.2.5"
 chrono = { version = "0.4.38", default-features = false, features = ["clock"] }
+chrono-tz = "0.4"
```
Comment (RalfJung) [inline Cargo.toml:28]: Why did you add such an ancient version of chrono-tz? The current version is 0.9.

File: src/shims/time.rs [hunk 3]
```diff
@@ -136,8 +138,16 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
             .unwrap();
         let dt_utc: DateTime<Utc> =
             DateTime::from_timestamp(sec_since_epoch, 0).expect("Invalid timestamp");
+
+        // Figure out what time zone is in use
+        let tz = this.get_var(OsStr::new("TZ"))?.unwrap_or_else(|| OsString::from("UTC"));
+        let tz = match tz.into_string() {
+            Ok(tz) => Tz::from_str(&tz).unwrap_or(Tz::Zulu),
+            _ => Tz::Zulu,
```
Comment (RalfJung) [inline src/shims/time.rs:146]: Why is Zulu the fallback, not UTC...?

File: src/shims/time.rs [hunk 4]
```diff
@@ -146,17 +156,17 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
 
         // tm_zone represents the timezone value in the form of: +0730, +08, -0730 or -08.
         // This may not be consistent with libc::localtime_r's result.
-        let offset_in_second = Local::now().offset().local_minus_utc();
-        let tm_gmtoff = offset_in_second;
+        let offset_in_seconds = Utc::now().with_timezone(&tz).offset().fix().local_minus_utc();
```
Comment (RalfJung) [inline src/shims/time.rs:159]: Can't `dt` be used for this, why do we need `Utc::now`?

File: Cargo.toml [hunk 5]
```diff
@@ -25,6 +25,7 @@ aes = { version = "0.8.3", features = ["hazmat"] }
 measureme = "11"
 ctrlc = "3.2.5"
 chrono = { version = "0.4.38", default-features = false, features = ["clock"] }
```
Comment (RalfJung) [inline Cargo.toml:27]: If I understand the docs correctly, we don't need the `clock` feature any more?

File: Cargo.toml (hunk 2 above)
Comment (saethlin) [reply Cargo.toml:28]: Probably because 4 looks like 9.

File: src/shims/unix/env.rs [hunk 6]
```diff
@@ -70,6 +70,27 @@ impl<'tcx> UnixEnvVars<'tcx> {
...
+        let _vars_ptr = ecx.read_pointer(&self.environ)?;
+        let Some(var_ptr) = self.map.get(name) else {
+            return Ok(None);
+        };
+        // The offset is used to strip the "{name}=" part of the string.
+        let var_ptr = var_ptr.offset(
+            Size::from_bytes(u64::try_from(name.len()).unwrap().checked_add(1).unwrap()),
+            ecx,
+        )?;
+        ecx.read_os_str_from_c_str(var_ptr).map(|s| Some(s.to_owned()))
```
Comment (RalfJung) [inline src/shims/unix/env.rs:92]: This duplicates a bunch of code in `getenv`, doesn't it? Can the code be shared, e.g. by having getenv call this?

File: src/shims/env.rs [hunk 7]
```diff
@@ -99,4 +99,15 @@ impl<'tcx> EnvVars<'tcx> {
 }
 
 impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriInterpCx<'mir, 'tcx> {}
-pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
+    /// Try to get an environment variable from the interpreted program's environment. This is
+    /// useful for implementing shims which are documented to read from the environment.
+    fn get_var(&mut self, name: &OsStr) -> InterpResult<'tcx, Option<OsString>> {
```
Comment (RalfJung) [inline src/shims/env.rs:105]: `get_env_var` may be better since this is a global namespace.

File: src/shims/unix/env.rs (hunk 6 above)
Comment (saethlin) [reply src/shims/unix/env.rs]: It does duplicate some code. `getenv` can't call it as-written but with some rearrangement it's possible to share the logic.
